tw/pkg/commands/dgrep/dgrep.go

```diff
@@ -0,0 +1,213 @@
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/chainguard-dev/clog"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/mattn/go-isatty"
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const (
+	DefaultTimeout = 5 * time.Second
+)
+
+type cfg struct {
+	Container   string
+	Timeout     time.Duration
+	IgnoreCase  bool
+	Retry       int
+	Patterns    []string
+	InvertMatch bool
+	Since       string
+	Tail        string
+
+	compiled    []*regexp.Regexp
+	highlighter func(string) string
+}
+
+func Command() *cobra.Command {
+	cfg := &cfg{}
+
+	cmd := &cobra.Command{
+		Use:          "dgrep CONTAINER [PATTERN]",
+		Short:        "Simple docker container log grepping",
+		Args:         cobra.MinimumNArgs(1),
+		SilenceUsage: true,
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			return cfg.prerun(cmd.Context(), args)
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return cfg.Run(cmd)
+		},
+	}
+
+	cmd.Flags().DurationVarP(&cfg.Timeout, "timeout", "t", DefaultTimeout, "time to wait for logs to appear")
+	cmd.Flags().IntVarP(&cfg.Retry, "retry", "r", 0, "number of times to retry a failed request")
+	cmd.Flags().BoolVarP(&cfg.IgnoreCase, "ignore-case", "i", false, "toggle to ignore case for the match")
+	cmd.Flags().StringArrayVarP(&cfg.Patterns, "regexp", "e", nil, "regular expression to match")
+	cmd.Flags().BoolVarP(&cfg.InvertMatch, "invert-match", "v", false, "toggle to invert the match")
+	cmd.Flags().StringVar(&cfg.Since, "since", "", "show logs since timestamp (e.g. 2013-01-02T13:23:37Z) or relative (e.g. 42m for 42 minutes)")
+	cmd.Flags().StringVar(&cfg.Tail, "tail", "", "number of lines to show from the end of the logs")
+
+	return cmd
+}
+
+func (c *cfg) Run(cmd *cobra.Command) error {
+	ctx := cmd.Context()
+
+	l := clog.FromContext(ctx).With("container", c.Container)
+
+	attempt := 0
+	err := wait.ExponentialBackoffWithContext(ctx, wait.Backoff{
+		Steps:    c.Retry + 1,
+		Duration: c.Timeout,
+		Factor:   1.0, // Keep backoff linear
+	}, func(ctx context.Context) (bool, error) {
+		attempt++
+
+		if err := c.retryableRun(ctx); err != nil {
+			l.ErrorContextf(ctx, "[%d/%d] failed to run dgrep: %v", attempt, c.Retry+1, err)
+			return false, nil
+		}
+
+		l.InfoContext(ctx, "dgrep succeeded", "attempt", attempt, "timeout", c.Timeout)
+		return true, nil
+	})
+	if err != nil {
+		return fmt.Errorf("dgrep failed after %d attempt(s)", attempt)
+	}
+
+	return nil
+}
+
+func (c *cfg) retryableRun(ctx context.Context) error {
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return fmt.Errorf("failed to create docker client: %v", err)
+	}
+	defer cli.Close()
+
+	options := container.LogsOptions{
+		ShowStdout: true,
+		ShowStderr: true,
+		Follow:     false,
+		Timestamps: false,
```

Comment:
nit: Enabling Timestamps would better IMO.
```suggestion
		Timestamps: true,
```